Feature: Allow stations under bridges and bridges over stations.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

JGRPP supports stations-under-bridges, vanilla does not.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Implement support for stations under bridges. Based on work by @Moth-Tolias which is based on the work in JGRPP. Although this is based on the previous work, it has many small changes and is structured quite differently.

Adds support for different station pieces to have different heights and bridge pillars, which allows for stations to fit under appropriately high bridges. Implements NewGRF variable support for bridges, stations and roadstops to specify these properties, compatible with JGRPP where possible.

Supersedes #14425.

<img width="1159" height="707" alt="image" src="https://github.com/user-attachments/assets/5266a5a5-378a-42d1-83cb-32b11b2b9346" />


<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

2TallTyler commented on src/clear_cmd.cpp:
```diff
@@ -390,6 +390,11 @@ static CommandCost TerraformTile_Clear(TileIndex tile, DoCommandFlags flags, int
+{
+	return CommandCost();
+}
+
```
This could use a comment explaining that we can always build above clear tiles.

2TallTyler commented on src/newgrf/newgrf_act0_bridges.cpp:
```diff
@@ -120,6 +121,20 @@ static ChangeInfoResult BridgeChangeInfo(uint first, uint last, int prop, ByteRe
+				break;
+
+			case 0x15: // Pillars
+				for (uint j = 0; j != 6; ++j) {
```
Magic 6. 🙂

2TallTyler commented on src/tunnelbridge_cmd.cpp:
```diff
@@ -1547,11 +1547,26 @@ static BridgePieces CalcBridgePiece(uint north, uint south)
 	}
 }
 
+BridgePillarFlags GetBridgeTilePillarFlags(TileIndex tile, TileIndex rampnorth, TileIndex rampsouth, BridgeType type, TransportType transport_type)
```
Missing doxygen.

2TallTyler commented on src/tunnelbridge_cmd.cpp:
```diff
@@ -1547,11 +1547,26 @@ static BridgePieces CalcBridgePiece(uint north, uint south)
  * @param ti Tile information of the tile to draw it on.
  */
-void DrawBridgeMiddle(const TileInfo *ti)
+void DrawBridgeMiddle(const TileInfo *ti, BridgePillarFlags blocked_pillars)
```
Doxygen params need updating.

2TallTyler commented on src/newgrf/newgrf_act0_roadstops.cpp:
```diff
@@ -134,6 +141,18 @@ static ChangeInfoResult RoadStopChangeInfo(uint first, uint last, int prop, Byte
+				break;
+
+			case 0x14: // Disallow pillars.
+				for (uint j = 0; j != 6; ++j) {
```
More magic 6 numbers. 🙂

2TallTyler commented on src/newgrf/newgrf_act0_stations.cpp:
```diff
@@ -268,11 +268,11 @@ static ChangeInfoResult StationChangeInfo(uint first, uint last, int prop, ByteR
-				buf.ReadWord();
+			case 0x1B: // Minimum bridge height (old variable)
+				if (statspec->tilespecs.size() < 8) statspec->tilespecs.resize(8);
+				for (int j = 0; j < 8; ++j) {
```
Magic 8, maybe even just naming this locally would be more clear.

2TallTyler commented on src/newgrf_station.h:
```diff
@@ -171,6 +172,8 @@ struct StationSpec : NewGRFSpecBase<StationClassID> {
 
 	struct TileSpec {
 		TileFlags flags{}; ///< Tile flags.
+		uint8_t height = 0; ///< Minimum height for a bridge above, 0 for none
```
None meaning no bridges allowed, or no restrictions for a bridge?

2TallTyler commented on src/saveload/saveload.h:
```diff
@@ -407,6 +407,7 @@ enum SaveLoadVersion : uint16_t {
 	SLV_INDUSTRY_NUM_VALID_HISTORY,         ///< 356  PR#14416 Store number of valid history records for industries.
 	SLV_INDUSTRY_ACCEPTED_HISTORY,          ///< 357  PR#14321 Add per-industry history of cargo delivered and waiting.
 	SLV_TOWN_SUPPLY_HISTORY,                ///< 358  PR#14461 Town supply history.
+	SLV_STATIONS_UNDER_BRIDGES,             ///< 359  PR#14477 Allow stations under bridges.
```
This savegame version doesn't seem to be used anywhere in the PR. Am I missing it?

PeterN replied:
No conversion is necessary. The savegame bump is to prevent loading a savegame with bridges over stations into a version of the game that doesn't support it.

LordAro replied:
maybe worth extending the comment?

2TallTyler commented on src/station_cmd.cpp:
```diff
@@ -864,6 +866,79 @@ static CommandCost CheckFlatLandAirport(AirportTileTableIterator tile_iter, DoCo
 	return cost;
 }
 
+static StringID GetDemolishStringForStationType(StationType type)
```
Missing doxygen.

PeterN replied:
Removed in favour of using the standard technique of trying clear the land, which will produce the correct error.

2TallTyler commented on src/station_cmd.cpp:
```diff
@@ -962,14 +1092,17 @@ static CommandCost CheckFlatLandRailStation(TileIndex tile_cur, TileIndex north_
  * @return The cost in case of success, or an error code if it failed.
  */
-static CommandCost CheckFlatLandRoadStop(TileIndex cur_tile, int &allowed_z, DoCommandFlags flags, DiagDirections invalid_dirs, bool is_drive_through, StationType station_type, Axis axis, StationID *station, RoadType rt)
+static CommandCost CheckFlatLandRoadStop(TileIndex cur_tile, int &allowed_z, const RoadStopSpec *spec, DoCommandFlags flags, DiagDirections invalid_dirs, bool is_drive_through, StationType station_type, Axis axis, StationID *station, RoadType rt)
```
Doxygen missing new parameter.

2TallTyler commented on src/station_cmd.cpp:
```diff
@@ -1392,8 +1528,29 @@ CommandCost CmdBuildRailStation(DoCommandFlags flags, TileIndex tile_org, RailTy
+	std::vector<uint8_t> layouts(numtracks * plat_len);
+	GetStationLayout(layouts.data(), numtracks, plat_len, statspec);
+
+	{
```
Doing something fancy with scope, or orphaned braces from a rebase gone wrong?

If the former, could use a comment explaining what this section does.

2TallTyler commented on src/station_cmd.cpp:
```diff
@@ -1913,7 +2065,7 @@ static CommandCost FindJoiningRoadStop(StationID existing_stop, StationID statio
  * @return The cost in case of success, or an error code if it failed.
  */
-CommandCost CalculateRoadStopCost(TileArea tile_area, DoCommandFlags flags, bool is_drive_through, StationType station_type, Axis axis, DiagDirection ddir, StationID *est, RoadType rt, Money unit_cost)
+CommandCost CalculateRoadStopCost(TileArea tile_area, DoCommandFlags flags, bool is_drive_through, StationType station_type, const RoadStopSpec *roadstopspec, Axis axis, DiagDirection ddir, StationID *est, RoadType rt, Money unit_cost)
```
Doxygen missing new param.

2TallTyler commented on src/station_cmd.cpp:
```diff
@@ -3073,6 +3225,20 @@ bool SplitGroundSpriteForOverlay(const TileInfo *ti, SpriteID *ground, RailTrack
+	if (layout < std::size(tilespecs)) return tilespecs[layout].disallowed_pillars;
+	return BRIDGEPILLARFLAGS_ALL;
+}
+
```
Missing doxygen.

2TallTyler commented on src/station_cmd.cpp:
```diff
@@ -5163,6 +5332,28 @@ uint FlowStatMap::GetFlowFromVia(StationID from, StationID via) const
 	return i->second.GetShare(via);
 }
 
+static CommandCost CheckBuildAbove_Station(TileIndex tile, DoCommandFlags flags, Axis, int height)
```
Missing doxygen.

PeterN replied:
Not documenting this one because we don't generally document callback implementations, but I have now documented the callback function signature instead.

2TallTyler reviewed (commented):
Haven't tested in depth yet, but here's an initial read-through of the code. Very excited for this one. 😄

2TallTyler commented on src/lang/english.txt:
```diff
@@ -5259,6 +5259,11 @@ STR_ERROR_START_AND_END_MUST_BE_IN                              :{WHITE}Start an
+STR_ERROR_BRIDGE_TOO_LOW_FOR_STATION                            :{WHITE}Bridge is too low for station
+STR_ERROR_BRIDGE_TOO_LOW_FOR_ROADSTOP                           :{WHITE}Bridge is too low for road stop
+STR_ERROR_BRIDGE_TOO_LOW_FOR_BUOY                               :{WHITE}Bridge is too low for buoy
+STR_ERROR_BRIDGE_TOO_LOW_FOR_RAIL_WAYPOINT                      :{WHITE}Bridge is too low for railway waypoint
```
Current strings call this a `train waypoint`, `rail waypoint`, or `railway waypoint`. What is most correct? (I'm happy to PR a change to the other strings for consistency.)

PeterN replied:
`railway waypoint` was the first I found, that's all. It's a bit of a mouthful. I think I prefer `rail waypoint` as that fits in with `road waypoint`

2TallTyler commented on src/bridge.h:
```diff
@@ -52,6 +39,8 @@ struct BridgeSpec {
 	std::vector<std::vector<PalSpriteID>> sprite_table; ///< table of sprites for drawing the bridge
 	uint8_t flags;                         ///< bit 0 set: disable drawing of far pillars.
+	BridgeSpecCtrlFlags ctrl_flags{}; ///< control flags
+	std::array<std::array<BridgePillarFlags, AXIS_END>, NUM_BRIDGE_PIECES - 1> pillar_flags{}; ///< bridge pillar flags, 6 for each axis.
```
Is there a reason this is different than elsewhere?

```suggestion
	static_assert(BRIDGE_PIECE_HEAD == 6);
	std::array<std::array<BridgePillarFlags, AXIS_END>, BRIDGE_PIECE_HEAD> pillar_flags{}; ///< bridge pillar flags, 6 for each axis.
```

2TallTyler reviewed (dismissed):
A few last height questions:
* Why do bay stops require height 2, while drive-through stops only require 1? At least in OpenGFX2, the building sprites are the same, just rearranged.
* Rail waypoints need a height of 2, but road waypoints could be 1. Why combine these?

2TallTyler reviewed (dismissed):
I don't think we need a setting. Players who don't want to use bridges over stations can just... not use them. 🙂

LordAro commented on src/bridge.h:
```diff
@@ -13,29 +13,19 @@
 constexpr uint SPRITES_PER_BRIDGE_PIECE = 32; ///< Number of sprites there are per bridge piece.
 
-typedef uint BridgeType; ///< Bridge spec number.
+enum class BridgeSpecCtrlFlag : uint8_t{
```
```suggestion
enum class BridgeSpecCtrlFlag : uint8_t {
```

LordAro commented on src/bridge.h:
```diff
@@ -13,29 +13,19 @@
 constexpr uint SPRITES_PER_BRIDGE_PIECE = 32; ///< Number of sprites there are per bridge piece.
 
-typedef uint BridgeType; ///< Bridge spec number.
+enum class BridgeSpecCtrlFlag : uint8_t{
```
doesn't this (and the others) belong in bridge_type.h ?

PeterN replied:
Potentially but `bridge_type.h` was created 1) to avoid a circular dependency (although that's no longer relevant due to redesign) and 2) to reduce a number of include dependencies. I've moved `BridgeSpecCtrlFlag` to be an internal enum of `BridgeSpec` now, which makes it a bit more contained.

LordAro commented on src/newgrf/newgrf_act0_bridges.cpp:
```diff
@@ -120,6 +121,22 @@ static ChangeInfoResult BridgeChangeInfo(uint first, uint last, int prop, ByteRe
 				bridge->price = buf.ReadWord();
 				break;
 
+			case 0x14: // Purchase sprite in JGRPP.
```
Why is this one explicitly being ignored? We don't ignore any of the other extra properties in JGRPP

(expand comment?) ("not yet supported"? )

PeterN replied:
Ignoring isn't exactly new. We previously ignored property 0x1B of stations. (And supporting that now is kinda pointless tbh.)

EDIT: Removing this one.

LordAro commented on src/newgrf_roadstop.h:
```diff
@@ -138,12 +139,10 @@ struct RoadStopSpec : NewGRFSpecBase<RoadStopClassID> {
 	uint8_t build_cost_multiplier = 16;  ///< Build cost multiplier per tile.
 	uint8_t clear_cost_multiplier = 16;  ///< Clear cost multiplier per tile.
 
+	std::array<BridgeableTileInfo, 6> bridgeable_info{}; ///< Per tile layout bridge information.
```
should probably try to avoid extra uses of literal '6'. NUM_BRIDGE_MIDDLE_PIECES  ?

PeterN replied:
Hmm, I don't think there's an existing constant for this. It's nothing to do with bridge pieces, it's the 6 possible orientations of road stops (NE, SE, SW, NW, X and Y)

LordAro commented on src/tunnelbridge_cmd.cpp:
```diff
@@ -1574,11 +1554,36 @@ static BridgePieces CalcBridgePiece(uint north, uint south)
+		BridgePieces piece = CalcBridgePiece(GetTunnelBridgeLength(tile, rampnorth) + 1, GetTunnelBridgeLength(tile, rampsouth) + 1);
+		Axis axis = TileX(rampnorth) == TileX(rampsouth) ? AXIS_Y : AXIS_X;
+
+		return spec->pillar_flags[piece][axis == AXIS_Y ? 1 : 0];
```
```suggestion
		return spec->pillar_flags[piece][TileX(rampnorth) == TileX(rampsouth)];
```

I'm not sure this actually improves things, but the intermediate Axis variable that's just used as a boolean flag feels a bit odd

PeterN replied:
The current code makes it clearer that we are interested in the axis of the bridge. Your suggestion seems to hide that.

LordAro commented on src/waypoint_cmd.cpp:
```diff
@@ -174,15 +174,15 @@ static CommandCost IsValidTileForWaypoint(TileIndex tile, Axis axis, StationID *
 	return CommandCost();
 }
 
 extern void GetStationLayout(uint8_t *layout, uint numtracks, uint plat_len, const StationSpec *statspec);
```
future work, make `layout` a proper return value?

PeterN replied:
"Future work" :-)

(But not really relevant to a review is it...)

PeterN committed:
Add: [NewGRF] Bridge pillar exclusion information for stations and roadstops.

NewGRF stations and roadstops can now specify bridge pillar and minimum height information for their tiles.

PeterN committed:
Codechange: Implement tile proc handler to test for building bridge over tile.

PeterN committed:
Change: Tweak tubular bridges to remove baked-in pillars.

PeterN committed:
Feature: Allow stations and roadstops under bridges.

Bridges above stations will have pillars excluded if they conflict with the station layout.

Partly based on the system implemented in JGRPP.

Co-authored-by: <su@angel-island.zone>

2TallTyler reviewed (approved):
Test in beta?
